common/src/main/java/com/wynntils/models/abilitytree/AbilityTreeContainerQueries.java

```diff
@@ -48,59 +50,68 @@ public void updateParsedAbilityTree() {
-            queryBuilder
-                    .clickOnSlotIfExists(PREVIOUS_PAGE_SLOT, DUMMY_SLOT)
-                    .expectSameContainer()
-                    .setAcceptNoOp(true)
-                    .processContainer(c -> {});
-        }
-
-        // endregion
-
-        // We are on the first page now. Re-read it by clicking a dummy slot,
-        // and accepting no-op changes.
-        queryBuilder
-                .clickOnSlot(DUMMY_SLOT)
-                .setAcceptNoOp(true)
-                .expectSameContainer()
-                .processContainer(c -> processor.processPage(c, 1));
-
-        for (int i = 2; i <= Models.AbilityTree.ABILITY_TREE_PAGES; i++) {
-            final int page = i; // Lambdas need final variables
-            queryBuilder
-                    .clickOnSlotWithName(NEXT_PAGE_SLOT, Items.STONE_AXE, NEXT_PAGE_ITEM_NAME)
-                    .expectSameContainer()
-                    .processContainer(c -> processor.processPage(c, page));
-        }
 
-        queryBuilder.build().executeQuery();
+                // Open character/compass menu
+                .then(QueryStep.useItemInHotbar(InventoryUtils.COMPASS_SLOT_NUM).expectContainerTitle("Character Info"))
+
+                // Open ability menu
+                .then(QueryStep.clickOnSlot(ABILITY_TREE_SLOT)
+                        .expectContainerTitle(Models.Container.ABILITY_TREE_PATTERN.pattern()))
+
+                // Go to first page, and save current page number
+                .execute(() -> this.pageCount = 0)
+                .repeat(
+                        c -> ScriptedContainerQuery.containerHasSlot(
+                                c, PREVIOUS_PAGE_SLOT, Items.STONE_AXE, PREVIOUS_PAGE_ITEM_NAME),
+                        QueryStep.clickOnSlot(PREVIOUS_PAGE_SLOT).processIncomingContainer(c -> {
+                            // Count how many times this is done, and save this value.
+                            // If we did not even enter here, we were already on first page.
+                            this.pageCount++;
+                        }))
+
+                // Process first page
+                .reprocess(processor::processPage)
+
+                // Repeatedly go to next page, if any, and process it
+                .repeat(
+                        c -> ScriptedContainerQuery.containerHasSlot(
+                                c, NEXT_PAGE_SLOT, Items.STONE_AXE, NEXT_PAGE_ITEM_NAME),
+                        QueryStep.clickOnSlot(NEXT_PAGE_SLOT).processIncomingContainer(processor::processPage))
+
+                // Go back to initial page
+                .repeat(
+                        c -> {
+                            // Go back as many pages as the original count was from the end
+                            this.pageCount++;
+                            return this.pageCount != Models.AbilityTree.ABILITY_TREE_PAGES;
+                        },
+                        QueryStep.clickOnSlot(PREVIOUS_PAGE_SLOT))
```

Comment:
This "resetting" is new. Is this needed / does this slow queries down? I assume this is more important in the quest book filters...

Reply:
Yes, it is new. I wouldn't say it slows down things much. In the worst case, you had the tree on the first page, then you need to go back 6 pages. My rough estimate is that 20 pages (the entire content book) takes 3-4 seconds, so at most an additional second.

I think we have to include your ability tree stuff for real to know if this becomes a problem. But I think it is proper of us to restore everything to the way it was, so our inspections are "hidden" for the user. So unless it is a serious problem I'd rather keep it. My thinking is that the reason this was not implemented before was that it was virtually impossible to do with the old design.